File: shared/thread-snooze.ts
// Per-thread snooze — one conversation put to sleep, not the whole bot.
//
// BotFleet already snoozes a WHOLE bot (`server/routines.ts`): that stops
// routines and webhooks restarting it, and it is what "snooze bot on stop"
// (#514) leaves behind.  This is the narrower thing — one thread of one bot
// goes quiet while its siblings keep working.
//
// The two compose in one direction only.  A snoozed bot implies every thread
// under it, because nothing may wake that bot at all; waking one thread never
// wakes the bot, because the person who stopped a bot did not ask for it back.
//
// Everything here is pure and shared, so the harness (`server/store.ts`,
// `server/index.ts`, `server/notify.ts`), the sidebar (`src/components/
// Sidebar.tsx`) and the paired phone's contract all answer "is this asleep?"
// the same way, from the same clock reading.

/** Sleep until the thread does something again rather than until a moment.
 *
 * 0 is a real value on the wire, never "no snooze": absent is how awake
 * travels, which is why the API takes JSON `null` to wake a thread and
 * treats an omitted field as "leave it alone". */
export const SNOOZE_UNTIL_ACTIVITY = 0;

/** Asleep right now.  The sentinel sleeps until something wakes it; a
 * timestamp sleeps only until it passes.
 *
 * The harness drops expired snoozes from what it hands out, but a live SSE
 * frame is never refreshed afterwards, so a client that has been open across
 * the deadline has to check the clock too. */
export function isThreadSnoozed(snoozedUntil: number | undefined, now = Date.now()): boolean {
  if (snoozedUntil === undefined) return false;
  return snoozedUntil === SNOOZE_UNTIL_ACTIVITY || snoozedUntil > now;
}

/** Whether a stored snooze has nothing left to wait for, so the harness may
 * drop it.  The sentinel is not a time and never expires on a clock. */
export function isSnoozeExpired(snoozedUntil: number | undefined, now = Date.now()): boolean {
  return snoozedUntil !== undefined && snoozedUntil > SNOOZE_UNTIL_ACTIVITY && snoozedUntil <= now;
}

/** The soonest still-future deadline in a list, or undefined when nothing is
 * scheduled to wake.  A list ticks its own clock off this: the sentinel never
 * ticks, and a deadline already past has nothing left to schedule. */
export function nextSnoozeExpiry(
  snoozes: readonly (number | undefined)[],
  now = Date.now(),
): number | undefined {
  let soonest: number | undefined;
  for (const until of snoozes) {
    if (until === undefined || until <= SNOOZE_UNTIL_ACTIVITY || until <= now) continue;
    if (soonest === undefined || until < soonest) soonest = until;
  }
  return soonest;
}

/** The presets the menu offers, in the order it offers them.  "custom" is
 * the person's own moment and resolves to whatever they picked. */
export const SNOOZE_PRESETS = ["hour", "tomorrow", "activity"] as const;
export type SnoozePreset = (typeof SNOOZE_PRESETS)[number];

/** Sentence case, because these are values in a menu rather than headings. */
export const SNOOZE_PRESET_LABELS: Record<SnoozePreset, string> = {
  hour: "For 1 hour",
  tomorrow: "Until tomorrow morning",
  activity: "Until activity",
};

/** Tomorrow morning is 8 local.  Local on purpose: it is the person's
 * morning, and the harness stores the absolute moment either way. */
export const SNOOZE_MORNING_HOUR = 8;

/** Resolve a preset when the person picks it, never when the menu opened —
 * a menu left open overnight must not snooze until a morning that has
 * already been and gone. */
export function resolveSnoozePreset(preset: SnoozePreset, now = Date.now()): number {
  if (preset === "activity") return SNOOZE_UNTIL_ACTIVITY;
  if (preset === "hour") return now + 3_600_000;
  const when = new Date(now);
  when.setDate(when.getDate() + 1);
  when.setHours(SNOOZE_MORNING_HOUR, 0, 0, 0);
  return when.getTime();
}

/** What a custom "wake me at" input means, or undefined when it means
 * nothing.  A moment already past is not a snooze, and neither is a string
 * the date parser could not read — both come back undefined so the caller
 * refuses rather than storing a snooze that is over before it starts. */
export function resolveCustomSnooze(value: string, now = Date.now()): number | undefined {
  const at = new Date(value).getTime();
  return Number.isFinite(at) && at > now ? at : undefined;
}

/** The one line a snoozed row says about itself: "until activity", or the
 * resolved deadline as "until 8:00 am".  Null when the thread is awake, so
 * a caller can render the badge or not on this alone.
 *
 * A deadline past today carries its weekday, because "until 8:00 am" on a
 * Friday row that wakes on Monday would be a lie. */
export function snoozeLabel(snoozedUntil: number | undefined, now = Date.now()): string | null {
  if (!isThreadSnoozed(snoozedUntil, now)) return null;
  if (snoozedUntil === SNOOZE_UNTIL_ACTIVITY) return "until activity";
  const when = new Date(snoozedUntil!);
  const time = when.toLocaleTimeString([], { hour: "numeric", minute: "2-digit" }).toLowerCase();
  const sameDay = when.toDateString() === new Date(now).toDateString();
  return sameDay ? `until ${time}` : `until ${when.toLocaleDateString([], { weekday: "short" })} ${time}`;
}

/** Newest activity, falling back to when the thread was created.  A missing
 * stamp sorts oldest rather than as NaN, so a half-loaded row cannot jump
 * the list. */
export function threadRecency(task: { createdAt: number; lastActivity?: number }): number {
  if (typeof task.lastActivity === "number" && Number.isFinite(task.lastActivity)) return task.lastActivity;
  return Number.isFinite(task.createdAt) ? task.createdAt : 0;
}

export interface SnoozeOrderedThread {
  threadId: string;
  createdAt: number;
  lastActivity?: number;
  snoozedUntil?: number;
}

/** Thread-list order: awake threads newest-first, then the sleeping ones.
 *
 * A snoozed thread SINKS rather than disappearing, and the moment its
 * deadline passes it is awake again and back in plain update order — nothing
 * has to remember where it used to sit, because the position was never
 * stored, only derived.
 *
 * `keepInPlace` is upstream #1248's pinned-thread retention, read literally:
 * an anchored thread is never demoted to the sleeping tier, so it holds the
 * update-order slot it already had rather than being promoted anywhere.
 * BotFleet pins bots rather than threads, so the anchor here is the thread
 * the person currently has open — snoozing the conversation you are reading
 * must not yank it out from under you.
 *
 * Stable within each tier: equal stamps keep the caller's order. */
export function orderedThreadRows<T extends SnoozeOrderedThread>(
  tasks: readonly T[],
  now = Date.now(),
  keepInPlace: ReadonlySet<string> = new Set(),
): T[] {
  const asleep = (task: T): boolean =>
    !keepInPlace.has(task.threadId) && isThreadSnoozed(task.snoozedUntil, now);
  return tasks
    .map((task, index) => ({ task, index, asleep: asleep(task) }))
    .sort((a, b) => {
      if (a.asleep !== b.asleep) return a.asleep ? 1 : -1;
      const recency = threadRecency(b.task) - threadRecency(a.task);
      return recency || a.index - b.index;
    })
    .map((entry) => entry.task);
}

File: ios/Sources/CompanionCore/ThreadSnooze.swift
// Per-thread snooze on the phone.
//
// One conversation goes quiet while the bot keeps working its others.  The
// harness owns the state (`shared/thread-snooze.ts`, `server/store.ts`) and
// this is the phone's half of the same contract, kept in the core rather
// than in a view so `swift test` can hold it to the desktop's answers.
//
// The wire is three values and nothing else: absent means awake, `0` is the
// until-activity sentinel and sleeps until the thread does anything again,
// and a timestamp in epoch milliseconds sleeps until that moment.  JSON
// `null` is the only way to wake a thread, because an omitted field has
// always meant "leave it alone" on the task route.
import Foundation

/// The sentinel: sleep until the thread does something again rather than
/// until a moment.  Zero is a real value on the wire, never "no snooze".
public let threadSnoozeUntilActivity: Double = 0

/// The snooze choices the phone offers, resolved in the person's local time
/// on purpose — it is their hour and their morning, and the harness stores
/// the absolute moment either way.
///
/// These mirror the desktop menu (`shared/thread-snooze.ts`) so a thread
/// snoozed on the phone reads the same on the Mac.
public enum ThreadSnoozePreset {
    /// Tomorrow morning is 8 local, matching the desktop.
    public static let morningHour = 8

    /// One hour from the tap.  Resolved when it is tapped, never when the
    /// menu was built: a menu left open must not snooze until a moment that
    /// has already been and gone.
    public static func hour(now: Date = Date()) -> Double {
        now.addingTimeInterval(3_600).timeIntervalSince1970 * 1_000
    }

    /// The next day at 8 local — a clean overnight break.
    public static func tomorrowMorning(now: Date = Date(), calendar: Calendar = .current) -> Double {
        let tomorrow = calendar.date(byAdding: .day, value: 1, to: now) ?? now
        let when = calendar.date(bySettingHour: morningHour, minute: 0, second: 0, of: tomorrow) ?? tomorrow
        return when.timeIntervalSince1970 * 1_000
    }
}

public enum ThreadSnooze {
    /// Asleep right now.  The sentinel sleeps until something wakes it; a
    /// timestamp sleeps only until it passes.
    ///
    /// The harness drops expired snoozes from what it hands out, so a fresh
    /// snapshot is authoritative — but a live SSE frame is never refreshed
    /// afterwards, and an app left open across a deadline has to check the
    /// clock too.
    public static func isSnoozed(_ snoozedUntil: Double?, now: Date = Date()) -> Bool {
        guard let snoozedUntil else { return false }
        return snoozedUntil == threadSnoozeUntilActivity || snoozedUntil > now.timeIntervalSince1970 * 1_000
    }

    /// The one line a sleeping row says about itself: "Until activity", or
    /// the resolved deadline as "Until 8:00 AM".  Nil while the thread is
    /// awake, so a caller can decide to draw the badge on this alone.
    ///
    /// A deadline past today carries its weekday, because "Until 8:00 AM" on
    /// a Friday row that wakes on Monday would be a lie.
    public static func label(
        _ snoozedUntil: Double?,
        now: Date = Date(),
        calendar: Calendar = .current,
        locale: Locale = .current
    ) -> String? {
        guard isSnoozed(snoozedUntil, now: now), let snoozedUntil else { return nil }
        if snoozedUntil == threadSnoozeUntilActivity { return "Until activity" }
        let when = Date(timeIntervalSince1970: snoozedUntil / 1_000)
        let time = DateFormatter()
        time.locale = locale
        time.calendar = calendar
        time.timeStyle = .short
        time.dateStyle = .none
        if calendar.isDate(when, inSameDayAs: now) { return "Until \(time.string(from: when))" }
        let weekday = DateFormatter()
        weekday.locale = locale
        weekday.calendar = calendar
        weekday.setLocalizedDateFormatFromTemplate("EEE")
        return "Until \(weekday.string(from: when)) \(time.string(from: when))"
    }

    /// Newest activity, falling back to when the thread was created, so a
    /// row that never spoke sorts by its own age rather than by nothing.
    public static func recency(_ task: BotTask) -> Double {
        if let lastActivity = task.lastActivity, lastActivity.isFinite { return lastActivity }
        return task.createdAt.isFinite ? task.createdAt : 0
    }

    /// Thread order: awake threads newest-first, then the sleeping ones.
    ///
    /// A snoozed thread SINKS rather than disappearing, and the moment its
    /// deadline passes it is awake again and back in plain update order —
    /// nothing has to remember where it used to sit, because the position
    /// was never stored, only derived.
    ///
    /// `keepInPlace` is upstream #1248's pinned-thread retention: an
    /// anchored thread is never demoted to the sleeping tier, so it holds
    /// the slot it already had.  BotFleet pins bots rather than threads, so
    /// the anchor here is the thread the person currently has open —
    /// snoozing the conversation on screen must not move it out from under
    /// the tap that snoozed it.
    ///
    /// Stable within each tier: equal stamps keep the caller's order.
    public static func ordered(
        _ tasks: [BotTask],
        now: Date = Date(),
        keepInPlace: Set<String> = []
    ) -> [BotTask] {
        let asleep = { (task: BotTask) -> Bool in
            !keepInPlace.contains(task.threadId) && isSnoozed(task.snoozedUntil, now: now)
        }
        return tasks.enumerated()
            .sorted { left, right in
                let leftAsleep = asleep(left.element)
                let rightAsleep = asleep(right.element)
                if leftAsleep != rightAsleep { return !leftAsleep }
                let leftAt = recency(left.element)
                let rightAt = recency(right.element)
                if leftAt != rightAt { return leftAt > rightAt }
                return left.offset < right.offset
            }
            .map(\.element)
    }
}
